Packet++/src/DoIpLayer.cpp

```diff
@@ -0,0 +1,1367 @@
+		case DoIpProtocolVersion::Version02Iso2012:
+		case DoIpProtocolVersion::Version03Iso2019:
+		case DoIpProtocolVersion::Version04Iso2019_AMD1:
+		case DoIpProtocolVersion::DefaultVersion:
+			return static_cast<DoIpProtocolVersion>(version);
+
+		default:
+			return DoIpProtocolVersion::UnknownVersion;
+		}
+	}
+
+	DoIpLayer::DoIpLayer()
+	{
+		m_DataLen = DOIP_HEADER_LEN;
+		m_Protocol = DOIP;
+		m_Data = new uint8_t[m_DataLen]{};
+	}
+
+	DoIpLayer::DoIpLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
+	    : Layer(data, dataLen, prevLayer, packet, DOIP)
+	{}
+
+	DoIpLayer* DoIpLayer::parseDoIpLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
+	{
+		doiphdr* doipHeader = reinterpret_cast<doiphdr*>(data);
+		uint16_t payloadType = doipHeader->payloadType;
+		DoIpPayloadTypes detectedPayloadType = static_cast<DoIpPayloadTypes>(htobe16(payloadType));
+
+		switch (detectedPayloadType)
+		{
+		case DoIpPayloadTypes::ROUTING_ACTIVATION_REQUEST:
+			return new RoutingActivationRequest(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::ROUTING_ACTIVATION_RESPONSE:
+			return new RoutingActivationResponse(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::GENERIC_HEADER_NEG_ACK:
+			return new GenericHeaderNack(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::VEHICLE_IDENTIFICATION_REQUEST_WITH_EID:
+			return new VehicleIdentificationRequestEID(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::VEHICLE_IDENTIFICATION_REQUEST_WITH_VIN:
+			return new VehicleIdentificationRequestVIN(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::ANNOUNCEMENT_MESSAGE:
+			return new VehicleAnnouncement(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::ALIVE_CHECK_RESPONSE:
+			return new AliveCheckResponse(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::DIAGNOSTIC_POWER_MODE_RESPONSE:
+			return new DiagnosticPowerModeResponse(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::ENTITY_STATUS_RESPONSE:
+			return new EntityStatusResponse(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::DIAGNOSTIC_MESSAGE_TYPE:
+			return new DiagnosticMessage(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::DIAGNOSTIC_MESSAGE_POS_ACK:
+			return new DiagnosticAckMessage(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::DIAGNOSTIC_MESSAGE_NEG_ACK:
+			return new DiagnosticNackMessage(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::VEHICLE_IDENTIFICATION_REQUEST:
+			return new VehicleIdentificationRequest(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::ALIVE_CHECK_REQUEST:
+			return new AliveCheckRequest(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::DIAGNOSTIC_POWER_MODE_REQUEST:
+			return new DiagnosticPowerModeRequest(data, dataLen, prevLayer, packet);
+		case DoIpPayloadTypes::ENTITY_STATUS_REQUEST:
+			return new EntityStatusRequest(data, dataLen, prevLayer, packet);
```

Comment:
Maybe all classe names should begin with `DoIp` so users know it belongs to this protocol?
For example:
```
RoutingActivationRequest  --> DoIpRoutingActivationRequest
RoutingActivationResponse --> DoIpRoutingActivationResponse
DiagnosticNackMessage     --> DoIpDiagnosticNackMessage
...
```